UefiPayloadPkg/Library/FdtParserLib/FdtParserLib.c

```diff
@@ -0,0 +1,627 @@
+              SmbiosTable->Header.Revision  = UNIVERSAL_PAYLOAD_SMBIOS_TABLE_REVISION;
+              SmbiosTable->Header.Length    = sizeof (UNIVERSAL_PAYLOAD_SMBIOS_TABLE);
+              SmbiosTable->SmBiosEntryPoint = (EFI_PHYSICAL_ADDRESS)(UINTN)(&StartAddress);
+            }
+          }
+        }
+      }
+    } // end of reserved-memory
+
+    if (AsciiStrnCmp (NodePtr->Name, "serial@", AsciiStrLen ("serial@")) == 0) {
+      //
+      // Create SerialPortInfo HOB.
+      //
+      Serial = BuildGuidHob (&gUniversalPayloadSerialPortInfoGuid, sizeof (UNIVERSAL_PAYLOAD_SERIAL_PORT_INFO));
+      ASSERT (Serial != NULL);
+      if (Serial == NULL) {
+        break;
+      }
+
+      Serial->Header.Revision = UNIVERSAL_PAYLOAD_SERIAL_PORT_INFO_REVISION;
+      Serial->Header.Length   = sizeof (UNIVERSAL_PAYLOAD_SERIAL_PORT_INFO);
+      Serial->RegisterStride  = 1;
+
+      for (Property = FdtFirstPropertyOffset (Fdt, Node); Property >= 0; Property = FdtNextPropertyOffset (Fdt, Property)) {
+        PropertyPtr = FdtGetPropertyByOffset (Fdt, Property, &TempLen);
+        TempStr     = FdtGetString (Fdt, Fdt32ToCpu (PropertyPtr->NameOffset), NULL);
+        if (AsciiStrCmp (TempStr, "current-speed") == 0) {
+          Data32 = (UINT32 *)(PropertyPtr->Data);
+          DEBUG ((DEBUG_INFO, "\n         Property(%08X)  %a", Property, TempStr));
+          DEBUG ((DEBUG_INFO, "  %X", Fdt32ToCpu (*Data32)));
+
+          Serial->BaudRate = Fdt32ToCpu (*Data32);
+        } else if (AsciiStrCmp (TempStr, "reg") == 0) {
+          Data32       = (UINT32 *)(PropertyPtr->Data);
+          Attribute    = Fdt32ToCpu (*(Data32 + 0));
+          StartAddress = Fdt32ToCpu (*(Data32 + 2));
+          DEBUG ((DEBUG_INFO, "\n         Property(%08X)  %a", Property, TempStr));
+          DEBUG ((DEBUG_INFO, " StartAddress   %016lX\n", StartAddress));
+          DEBUG ((DEBUG_INFO, " Attribute      %016lX\n", Attribute));
+
+          Serial->RegisterBase = StartAddress;
+          Serial->UseMmio      = (Attribute&SS_32BIT_MEMORY_SPACE) == SS_32BIT_MEMORY_SPACE ? TRUE : FALSE;
+        }
+      }
+    } else if (AsciiStrCmp (NodePtr->Name, "graphic") == 0) {
+      //
+      // Create GraphicInfo HOB.
+      //
+      GraphicsInfo = BuildGuidHob (&gEfiGraphicsInfoHobGuid, sizeof (EFI_PEI_GRAPHICS_INFO_HOB));
+      ASSERT (GraphicsInfo != NULL);
+      if (GraphicsInfo == NULL) {
+        break;
+      }
+
+      GraphicsDev = BuildGuidHob (&gEfiGraphicsDeviceInfoHobGuid, sizeof (EFI_PEI_GRAPHICS_DEVICE_INFO_HOB));
+      ASSERT (GraphicsDev != NULL);
+      if (GraphicsDev == NULL) {
+        break;
+      }
+
+      if (FrameBufferBase != 0) {
+        GraphicsInfo->FrameBufferBase = FrameBufferBase;
+      }
+
+      if (FrameBufferSize != 0) {
+        GraphicsInfo->FrameBufferSize = (UINT32)FrameBufferSize;
+      }
+
+      for (Property = FdtFirstPropertyOffset (Fdt, Node); Property >= 0; Property = FdtNextPropertyOffset (Fdt, Property)) {

```

Comment:
Per UPL specification, 'graphic' node should be 'framebuffer' node, example as below:
1. Fill HOB frameBufferBase/Size by FDT framebuffer->reg

2. Fill HOB for X and Y:
  GraphicsInfo->GraphicsMode.HorizontalResolution = framebuffer->width 
  GraphicsInfo->GraphicsMode.VerticalResolution = framebuffer->height

3. use framebuffer->format property to fill PixelFormat:
  GraphicsInfo->GraphicsMode.PixelFormat : framebuffer->format:
    PixelRedGreenBlueReserved8BitPerColor : 'a8r8g8b8'
    PixelBlueGreenRedReserved8BitPerColor :  'a8b8g8r8'
    PixelFormatMax (Unsupported by current UefiPayload) : 'a16b16g16r16'

4. HOB PixelInformation (BitMask) currently is not used by UPL, you may keep them as 0.

5. Fill HOB PixelsPerScanLine
  GraphicsInfo->GraphicsMode.PixelsPerScanLine = framebuffer->stride

6. GraphicsDev HOB is for PCI GFX device only and all fields should be filled by "PCI GFX device node", which will be pointed by framebuffer->display. Below is example of PCI GFX node

rb@E0000000 {
      compatible = "pci-host-ecam-generic";
      reg = <...>
      ranges = <...>;
	gma@E0010000 {
                 /* gfx device 00:02:00 */
                vendor-id = <0x8086>
                device-id = <0x4680>
                revision-id = <0x4>
                subsystem-vendor-id = <0x8086>
                subsystem-id = <0x2212>
            };
      };
};

framebuffer@b0000000 {
    compatible = "simple-framebuffer";
    reg = <0x82000000, 0x0, 0xb0000000, 0x400, 0x500000>; //base 0xB0000000, size=0x4000500000
    width = <3840>;
    height = <2160>;
    format = "a8r8g8b8";
    display = <&gma>;
};

Note1: GraphicsDev HOB default is all 0xFF, please SetMem() of this buffer to 0xFF before parsing FDT nodes.
Note2: GraphicsDev->BarIndex is not part of standard DT PCI binding, so leave it as default 0xFF.

Reply:
resolved